esp/services/ws_dfu/ws_dfuService.cpp

```diff
@@ -6118,7 +6130,329 @@ bool CWsDfuEx::onDFUFileAccess(IEspContext &context, IEspDFUFileAccessRequest &r
+        StringBuffer jsonLayout;
+        Owned<IRtlFieldTypeDeserializer> deserializer(createRtlFieldTypeDeserializer());
+        const RtlTypeInfo *typeInfo = deserializer->deserialize(layoutBin);
+        dumpTypeInfo(jsonLayout, typeInfo);
+        if (jsonLayout.length())
+            accessInfo.setRecordTypeInfoJson(jsonLayout.str());
+    }
+    if (req.getReturnBinTypeInfo())
+    {
+        MemoryBuffer binLayout;
+        layoutBin.swapWith(binLayout);
+        if (binLayout.length())
+            accessInfo.setRecordTypeInfoBin(binLayout);
+    }
+}
+
+bool CWsDfuEx::onDFUFileCreate(IEspContext &context, IEspDFUFileCreateRequest &req, IEspDFUFileCreateResponse &resp)
+{
+    try
+    {
+        IConstDFUFileAccessRequestBase &requestBase = req.getRequestBase();
+        SecAccessFlags accessType = translateToSecAccessFlags(requestBase.getAccessType());
+        if (SecAccess_None == accessType)
+        {
+            context.setAuthStatus(AUTH_STATUS_NOACCESS);
+            throw MakeStringException(ECLWATCH_DFU_ACCESS_DENIED, "WsDfu::DFUFileCreate - Permission denied.");
+        }
+        context.ensureFeatureAccess(FEATURE_URL, accessType, ECLWATCH_DFU_ACCESS_DENIED, "WsDfu::DFUFileCreate: Permission denied.");
+
+        const char *fileName = requestBase.getName();
+        const char *clusterName = requestBase.getCluster();
+        const char *recordDefinition = req.getECLRecordDefinition();
+        if (isEmptyString(fileName))
+             throw MakeStringException(ECLWATCH_INVALID_INPUT, "No Name defined.");
+        if (isEmptyString(clusterName))
+             throw MakeStringException(ECLWATCH_INVALID_INPUT, "No Cluster defined.");
+        if (isEmptyString(recordDefinition))
+             throw MakeStringException(ECLWATCH_INVALID_INPUT, "No ECLRecordDefinition defined.");
+
+        Owned<IConstWUClusterInfo> clusterInfo = getTargetClusterInfo(clusterName);
+        if (!clusterInfo)
+             throw MakeStringException(ECLWATCH_INVALID_INPUT, "Cluster %s not found.", clusterName);
+
+        ClusterType clusterType = clusterInfo->getPlatform();
+        const char *clusterTypeEx = (clusterType == ThorLCRCluster) ? "thor" :
+            ((clusterType == RoxieCluster) ? "roxie" : "hthor");
+
+        StringBuffer groupName;
+        Owned<IGroup> group = queryDFUFileIGroup(clusterName, clusterType, clusterTypeEx, req.getPartLocations(), groupName);
+
+        MemoryBuffer layoutBin;
+        exportRecordDefinitionBinaryType(recordDefinition, layoutBin);
+
+        StringBuffer userId;
+        Owned<IUserDescriptor> userDesc;
+        context.getUserID(userId);
+        if (!userId.isEmpty())
+        {
+            userDesc.setown(createUserDescriptor());
+            userDesc->set(userId.str(), context.queryPassword(), context.querySignature());
+        }
+
+        StringBuffer tempFileName = fileName;
+        tempFileName.append(".wsdfucreate.tmp");
+
+        StringBuffer jobId = requestBase.getJobId();
```

Comment:
Unnecessary use of StringBuffer (and therefore copy).

Reply:
I use it because I need to use the appendf in the next two lines.

Reply:
ok, yes I missed that.